src/test/java/com/disney/pg2k4j/containers/KinesisLocalStack.java

```diff
@@ -0,0 +1,143 @@
+package com.disney.pg2k4j.containers;
+
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.services.kinesis.AmazonKinesis;
+import com.amazonaws.services.kinesis.AmazonKinesisClientBuilder;
+import com.amazonaws.services.kinesis.model.DescribeStreamRequest;
+import com.amazonaws.services.kinesis.model.GetRecordsRequest;
+import com.amazonaws.services.kinesis.model.GetRecordsResult;
+import com.amazonaws.services.kinesis.model.GetShardIteratorRequest;
+import com.amazonaws.services.kinesis.model.GetShardIteratorResult;
+import com.amazonaws.services.kinesis.model.ListShardsRequest;
+import com.amazonaws.services.kinesis.model.ListShardsResult;
+import com.amazonaws.services.kinesis.model.ShardIteratorType;
+import com.amazonaws.services.kinesis.model.StreamDescription;
+import com.amazonaws.waiters.FixedDelayStrategy;
+import com.amazonaws.waiters.MaxAttemptsRetryStrategy;
+import com.amazonaws.waiters.PollingStrategy;
+import com.amazonaws.waiters.WaiterParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.wait.strategy.LogMessageWaitStrategy;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+
+import static org.testcontainers.containers.BindMode.READ_WRITE;
+
+public class KinesisLocalStack<SELF extends GenericContainer<SELF>> extends
+        GenericContainer<SELF> {
+
+    private static final int KINESIS_PORT = 4568;
+    private static final String KINESIS_SERVICE_NAME = "kinesis";
+    public static final String STREAM_NAME = "postgres_cdc";
+    public static final int NUM_SHARDS = 1;
+    public AmazonKinesis client;
+
+    private static final Logger logger =
+            LoggerFactory.getLogger(KinesisLocalStack.class);
+
+    public KinesisLocalStack(final Network network) {
+        super("localstack/localstack:latest");
+
+        this.withNetwork(network)
```

Comment:
An alternative approach would be to create a localstack environment (like this: https://github.com/WW-Digital/reactive-kinesis/blob/master/localstack/docker-compose.yml) and have travis spawn it. Then run against this.

Reply:
this would definitely make the code cleaner, at least in my eyes - but I'm more familiar with that approach admittedly. What do you think about raising this as an issue for the future?

Reply:
We've been discussing internally about the direction we want to go with our integration tests. Currently our tests run with an [externalized docker compose file](https://github.bamtech.co/personalization/kirby/blob/master/integrationtests_resources/docker-compose-infra.yml), but recently we've decided to migrate to this test-containers solution. Defining all the infrastructure needed to run the test within the junit test itself makes the process of debugging easier (right-click debug in Intellij). I don't have a huge preference here so let me know if there's any pro or con that I'm not considering heavily enough.

Reply:
We've been back and forth on this quite a bit in the last year or two, and we consider this test-container approach to be the lowest barrier to entry solution with the largest ROI we've found yet. I think it's a large enough component to warrant a discussion here about it, so please don't take any of this as nitpicky :) We'd love to discuss the pros and cons with someone who knows about localstack and has fresh eyes.

The tl;dr is that we've found ourselves maintaining a lot of scripts to manage the spin-up of infrastructure, a lot of manual health check logic to wait for different services to be "ready" for test execution ([which doesn't work as out-of-the-box as we would hope](https://docs.docker.com/compose/startup-order/), and a web of complexity when dealing with local debugging of individual components (run the tests, see them fail, keep the services spun up while we execute changed code against it). Not to mention easily running the integration tests locally and consistently.

Further, the infrastructure is pretty far removed from the test contexts themselves. We run a fairly large distributed system, and it can be tricky to visualize the entire architecture while running the tests (and reason about it as a newcomer). 

Test-containers gives us a totally contained context to define what to spin up (and how), providing native solutions for waiting for services to initialize and come online. To me, it's a lot cleaner and comprehensible to see familiar JUnit tests run against some abstracted away infrastructure, trusting that the infra is managed as part of the test lifecycle itself. As @rkass mentioned, it's super nice to be able to right click -> run a test inside of your IDE.

That said, we're open to alternatives. We can definitely sync offline to do a deeper dive on this to understand our current pain points. We realize this project is just a sliver of our entire infra, and can probably more discretely be represented by docker-compose and travis-spawned test infra, but to me it creates a little more mystery to what's actually happening under the hood.

Reply:
Just realised I didn't respond to this (I read it on the email notification whilst travelling). 

This makes sense, I find it easier to reason about the env when it's defined separately (e.g. in docker-compose). This may also include other service dependencies. I created the localstack wrapper which adds additional healthcheck capabilities but I take your point with that.  I would normally have travis or sbt spin it up and then the app would run against this (so no scripts, kind of).

However I don't feel strongly about this and it seems like you've spent a lot of time getting a solution which works for you. It's not something which I feel requires consistency across the org. Certainly interested in hearing about your journey to this point though as any learnings are valuable!